File: source/javascript/file_watcher.js
// File watcher: write a file, check for changes, detect modification
const fs = require('fs');

const watchPath = '/tmp/js_watch_test.txt';

fs.writeFileSync(watchPath, 'initial content');
const t1 = Date.now();
console.log(`[watcher] monitoring: ${watchPath}`);
console.log(`[watcher] initial write at t=${t1}`);

const initial = fs.readFileSync(watchPath, 'utf8');
console.log(`[watcher] initial content: ${initial}`);

fs.writeFileSync(watchPath, 'modified content');
const t2 = Date.now();
console.log(`[watcher] file modified at t=${t2}`);

const content = fs.readFileSync(watchPath, 'utf8');
if (content === 'modified content') {
    console.log('[watcher] file change detected');
    console.log(`[watcher] new content: ${content}`);
} else {
    console.log('[watcher] no change detected');
}

if (t2 > t1) console.log('[watcher] modification time advanced');

fs.unlinkSync(watchPath);
console.log('[watcher] done');

File: source/go/file_watcher.go
package main

import (
	"fmt"
	"os"
	"time"
)

func main() {
	watchPath := "/tmp/hemlock_watch_test.txt"

	err := os.WriteFile(watchPath, []byte("initial content"), 0644)
	if err != nil {
		fmt.Println("write error:", err)
		return
	}
	t1 := time.Now().UnixMilli()
	fmt.Printf("[watcher] monitoring: %s\n", watchPath)
	fmt.Printf("[watcher] initial write at t=%d\n", t1)

	contents, _ := os.ReadFile(watchPath)
	fmt.Printf("[watcher] initial content: %s\n", string(contents))

	err = os.WriteFile(watchPath, []byte("modified content"), 0644)
	if err != nil {
		fmt.Println("write error:", err)
		return
	}
	t2 := time.Now().UnixMilli()
	fmt.Printf("[watcher] file modified at t=%d\n", t2)

	content, _ := os.ReadFile(watchPath)
	if string(content) == "modified content" {
		fmt.Println("[watcher] file change detected")
		fmt.Printf("[watcher] new content: %s\n", string(content))
	} else {
		fmt.Println("[watcher] no change detected")
	}

	if t2 > t1 {
		fmt.Println("[watcher] modification time advanced")
	}

	fmt.Println("[watcher] done")
}
